Ba 24 initialise a checkout procedure for a new customer

jakub-krzysiek opened the pull request:
### **NOTE: This is NOT working 100% as intended. This is mainly for Code Review and discussion.**

## Description:
This pull request introduces `LoginPage` class (Page Object Model) implementation and includes amendments to `BasePage`, `PropertyManager` and associated classes.

## Changes:
**`LoginPage` Page Object Model:**

1. Added the initial version of the `LoginPage` class, which extends the `BasePage` class.
2. Implemented methods for interacting with login elements such as email, password, login button, registration and guest checkout options.

**`BasePage` updates:**

1. Modified the constructor to initialize WebDriver based on the browser name.
2. Updated the wait time configuration to utilize a global wait value obtained from `PropertyManager`.
3. Adapted base URL retrieval to use the `PropertyManager`.

**`PropertyManager` modifications:**

1. Amended 'GLOBAL_WAIT' parameter's name for the global wait configuration.

dariabajda commented on src/main/java/pages/LoginPage.java:
```diff
@@ -0,0 +1,40 @@
+    // Locators:
+    private By loginInputLocator = By.xpath("//*[@id=\"loginFrm_loginname\"]");
+    private By passwordInputLocator = By.xpath("//*[@id=\"loginFrm_password\"]");
+    private By loginButtonLocator = By.xpath("//*[@id=\"loginFrm\"]/fieldset/button");
```
This locator might be unstable since it's assuming we have a strict structure in the code.
How about more loose one: `//*[@id=\"loginFrm\"]//button`?

dariabajda replied:
Also you can use single qoutes inside the selector: `By.xpath("//*[@id='loginFrm_password']")`

jakub-krzysiek replied:
@dariabajda After today's discussion, we agreed on simply choosing the best possible syntax. These locators have now been changed to By.id() / Changed double quoting to single quoting for the 'loginButtonLocator' (which is By.xpath)

dariabajda commented on src/main/java/pages/LoginPage.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    // Locators:
+    private By loginInputLocator = By.xpath("//*[@id=\"loginFrm_loginname\"]");
```
Why not simply `By.id`?

jakub-krzysiek replied:
@dariabajda Answered in the comment above :)

dariabajda commented:
In the future I would propose to create separate PRs out of these changes - page object changes and webdriver changes seem unrelated

jakub-krzysiek commented:
`LoginPage 2.0` commit:

1. WebDriver was being instantiated more than once. `WebDriverFactory` class has been amended to a singleton pattern.
2. `PropertyManager` class: as there are more than one instance 'checks' now, changed the `instance` variable name to `propertyManagerInstance`
3. `BasePage` class: tweaked Line 38 (methods' names in `WebDriverFactory` class have been slightly amended).
4. `LoginPage` class:
    a.  Changed the inheritance (by extending `BasePage` class) to a composition scheme - all methods now have a reference to 
         specific `BasePage` class methods, instead of inheriting everything
    b.  Added `if` clause to the `selectGuestAccountOption` method - if the cart is empty, the button does not show / if there are 
         items in cart, the button appears.

Please review again :)

Fishgame1 commented on src/main/java/pages/BasePage.java:
```diff
@@ -32,8 +34,8 @@ public class BasePage {
 
-    public BasePage(WebDriver driver) {
-        this.driver = driver;
+    public BasePage(BrowserName browserName) {
```
Maybe we can just take the browserName from the config instead of passing it here?

jakub-krzysiek replied:
@Fishgame1 This has now been done in the second latest commit

Fishgame1 commented on src/main/java/pages/LoginPage.java:
```diff
@@ -0,0 +1,45 @@
+    public LoginPage(BrowserName browserName) {
+    }
+
+    // Methods:
```
unnecessary comment.

jakub-krzysiek replied:
@Fishgame1 Removed

Fishgame1 commented on src/main/java/pages/LoginPage.java:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    public void selectGuestAccountOption() {
+        if (basePage.isElementDisplayed(guestCheckoutButtonLocator)) {
```
We don't need this exception in the tests. I would rather have method with parameter type boolean to check if this element is hidden or not and for it to be a part of some test.

jakub-krzysiek replied:
@Fishgame1 Method with boolean parameter added to new test 'TestGuestCheckout'.

Fishgame1 replied:
So can we remove that exception from here?:)

jakub-krzysiek replied:
@Fishgame1 As per previous comments, I think it'd be best to fix it as part of basic tests rework - not only would I have to remove the exception from `LoginPage` class, I'd also use the methods provided in `BasePage` class for this

Fishgame1 commented on src/test/java/temp/tests/TestOpenPageChrome.java:
```diff
@@ -1,25 +1,27 @@
     public void openPage() {
         getPropertyManagerInstance();
-        WebDriver driver = WebDriverFactory.getWebDriver(BrowserName.CHROME);
+        WebDriver driver = getWebDriverInstance(BrowserName.CHROME);
```
I would move web driver init to the basePage class or to some base test class - I think there was a possibility to do that so TestOpenPageChrome would extend some baseTest class which can do the initialization:)

jakub-krzysiek replied:
@Fishgame1 Done - web driver initialization is now being done only in BasePage

Fishgame1 replied:
Ok, so we can remove that line here right?:)

jakub-krzysiek replied:
@Fishgame1 Web driver init left only in BasePage class

Fishgame1 commented on .gitignore:
```diff
@@ -62,7 +62,6 @@ build/
 *.rar
 *.err
 
-temp/
```
why you removed this one?

Fishgame1 replied:
Oh...I see...it is kind of wrong to do this like that...maybe keep temp/ here and rename temp/tests to work.in.progress.tests?

jakub-krzysiek replied:
@Fishgame1 I was thinking about creating a new task for the basic tests rework - to be discussed tomorrow.

Fishgame1 commented on src/test/java/temp/tests/TestGuestCheckout.java:
```diff
@@ -0,0 +1,90 @@
+
+    @Test
+    public void goToCheckout() {
+        addToCartAndGoToCart();
```
nonononononono!!! We cannot make tests that depends one on another:) create class with steps and define them there to reuse them in the tests :) could you also add comment that some of the steps are dummy steps that will be implement later on? Maybe put those into some steps as well so it will be easy to change after the implementation is ready (you can allign with creators if any parameter is required there) :)

jakub-krzysiek replied:
@Fishgame1 Basic tests rework to be discussed tomorrow - as per my comment above

Fishgame1 commented on src/test/java/temp/tests/TestOpenPageFirefox.java:
```diff
@@ -18,7 +20,8 @@ public class TestOpenPageFirefox {
     public void openPage() {
         getPropertyManagerInstance();
-        WebDriver driver = WebDriverFactory.getWebDriver(BrowserName.FIREFOX);
+        WebDriver driver = getWebDriverInstance(BrowserName.CHROME);
```
remove that line :)

jakub-krzysiek replied:
@Fishgame1 Basic tests rework to be discussed tomorrow - as per my comment above

Fishgame1 commented on src/test/java/temp/tests/TestOpenPageOpera.java:
```diff
@@ -19,7 +20,8 @@ public class TestOpenPageOpera {
     public void openPage() {
         getPropertyManagerInstance();
-        WebDriver driver = WebDriverFactory.getWebDriver(BrowserName.OPERA);
+        WebDriver driver = getWebDriverInstance(BrowserName.CHROME);
```
you know what to do at this point :)

jakub-krzysiek commented:
I was fighting with this for far too long :D
New commit [5b67467] - New version of passing browser name directly from config file:
1. `PropertyManager` class:
  Instead of doing it in `BasePage` class, I thought it would be appropriate to retrieve the web browser name directly from config file within the class that handles config file - `getBrowserName` method was implemented,
2. `WebDriverFactory` class:
    Removed `BrowserName browserName` parameter from both `getWebDriverInstance` and `createWebDriverInstance` methods,
    `createWebDriverInstance` method's switch function now uses a new `getBrowserName` method,
3. `BasePage` class:
    Changed retrieving web browser name to just using the new `getBrowserName` method,
    Changed initializing web driver from declaring a specific browser to initializing it based on the `getBrowserName` method (which is used in a method that creates a new web driver instance in `WebDriverFactory` class),
4. Test classes:
    Removed directly declaring the web browser name.
